### ISSUE TITLE
Winforms Customer Designer - DataGridView

### ISSUE THREAD

**Issue by dmanter** [dotnet/winforms#8508]
### .NET version

7.0

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

Not tried - Worked in 4.7.2

### Issue description

Hi Klaus,
I can’t find anywhere or anybody to help me with this problem. I had a programmer who wrote us a custom Form Designer using some old VS Designer Project from many years ago. Basically we allow user to design a form and then save it and the controls to xml. At runtime we read the xml make a dynamic user form with those controls. When they hit OK, it runs a stored procedure with SQL and the controls text entries are passed to the stored procedure as parameters. This allow things like a user can make a pricing update form with a simple textbox for price increase amount. The stored proc runs thru the price table and adds that. We have been using for many years. Just updating now off .Net 4 to .Net 7 to modernize our app. This Form Designer almost worked. The DateTimePicker and DataGridView get a no designer error. I’ve spent days tracking the code that fail and only get to this in the TypeDescriptor.
The line :type = Type.GetType(designerAttribute.DesignerTypeName); returns null for DTP and DGV but a designer for Labels, TextBox, RadioBUttons, CHeckBOx, etc.
Only null on DTP and DGV.
I can’t seem to get debug to step into the Type.GetType() so at a dead end.
I am stuck for days and at a loss what to do. DTP and DGV work in VS 2022 designer so there must be some way this can get at a designer for each.
Thanks Doug Manter, CTO Creative Information Systems, Inc.
[dmanter@creativeinfo.net](mailto:dmanter@creativeinfo.net)
Please help or advise. I wanted to leave under WinForms page but said comments closed.

    public static IDesigner? CreateDesigner(IComponent component, Type designerBaseType)
    {
        Type? type = null;
        IDesigner? result = null;
        AttributeCollection attributes = GetAttributes(component);
        for (int i = 0; i < attributes.Count; i++)
        {
            if (attributes[i] is DesignerAttribute designerAttribute)
            {
                Type? type2 = Type.GetType(designerAttribute.DesignerBaseTypeName);
                if (type2 != null && type2 == designerBaseType)
                {
                    ISite? site = component.Site;
                    bool flag = false;
                    ITypeResolutionService? typeResolutionService = (ITypeResolutionService?)site?.GetService(typeof(ITypeResolutionService));
                    if (typeResolutionService != null)
                    {
                        flag = true;
                        type = typeResolutionService.GetType(designerAttribute.DesignerTypeName);
                    }
                    if (!flag)
                    {
                        type = Type.GetType(designerAttribute.DesignerTypeName); <- RETURNS NULL for DateTimePicker and DataGridview.
                    }
                    if (type != null)
                    {
                        break;
                    }
                }
            }
        }
        if (type != null)
        {
            result = (IDesigner?)Activator.CreateInstance(type);
        }
        return result;
    }
![CustomFormDesignerIssue](https://user-images.githubusercontent.com/3495842/212772762-218692b8-3d62-4aca-90ed-3e5e2c4696c2.png)

### Steps to reproduce

I could upload the whole project but unsure how.

**Comment by elachlan** [comment]
Related: #8501

**Comment by KlausLoeffelmann** [comment]
What @elachlan wrote is correct in so far as it is currently out of scope for our immediate goals.

I can take this to the next team meeting, but I don't think this has a chance to raise in priority any time soon.

**Comment by merriemcgaw** [comment]
@dmitrii-drobotov can your team take a look at porting these two designers?

**Comment by dmanter** [comment]
> @dmanter, I want to confirm that you are asking about designer support in the runtime of your application? We do have a VS WinForms Extensibility SDK that has those features in the Visual Studio environment. If you are looking for VS related support, we have both designers in the SDK and are working on a CustomRootDesigner to allow developers to create their own designers in the context of VS. But @elachlan is correct, we don't have runtime support for these designers yet as the customer requests have been very low. This is something we can port when there is enough demand.

Yes - My App "FormDesigner" is an EXE that at runtime, emulates the form designer in VS to allow end users to make Forms that when saved, we store into XML files. We have another app that reads the XML and dynamically builds a form with the controls the user dropped on it. I hope this answers you
![Capture](https://user-images.githubusercontent.com/3495842/213354921-1436f864-3ec8-45c1-baed-1864d2101937.JPG)
r question.

**Comment by Leon99** [comment]
```csharp
    internal class DateTimePickerDesigner : ControlDesigner
    {
        public DateTimePickerDesigner()
        {
            AutoResizeHandles = true;
        }

        public override SelectionRules SelectionRules => base.SelectionRules & ~(SelectionRules.TopSizeable | SelectionRules.BottomSizeable);
    }

    [Designer("AwesomeCompany.AwesomeProduct.DateTimePickerDesigner, AwesomeCompany.AwesomeProduct")]
    public class CustomDateTimePicker : DateTimePicker{}
```
Missing `DateTimePickerDesigner` is the only issue that we had. I imagine only code involved in proper implementation of `DateTimePickerDesigner.SnapLines` was not that straightforward to port to .NET 6, but since we don't care about snap lines that's enough. @dmanter you're welcome :)

**Comment by dmitrii-drobotov** [comment]
@merriemcgaw sure, we'll look into it. For `DateTimePicker` it should be pretty simple, `SnapLines` was already ported to  `ControlDesigner` and others, so we can make a small separate PR with it. But for `DataGridView` it's definitely going to more complex and I think we'd firstly need to estimate the work on it.

**Comment by dmanter** [comment]
> 

Hi @Leon99 - Worked!!!! Thanks. Now just need to get the DataGridView issue resolved! Many thanks!

**Comment by dmanter** [comment]
> > 
> 
> Hi @Leon99 - Worked!!!! Thanks. Now just need to get the DataGridView issue resolved! Many thanks!

OK - Now tried it for DataGridView the same way and that allowed me to get that working.
Only issue I have now is all the controls work and allow placement on form but I can't size them. When I hover over the sizing block I get no cursor change and or sizing action. You've helped me at least get these last two controls now onto my Form and I can size manually in the Property Editor but not with mouse. Ideas? I am very new to this forum and don't want to violate any rules but I really could use some consulting help in this .Net Core port from Framework. Any ideas! If this is wrong to ask, please forgive me... Is there another forum for that?
Thanks!
Doug

**Comment by dmitrii-drobotov** [comment]
@dmanter thank you for a report, this seems to be a bug in .NET 7. We have the [DemoConsole](https://github.com/dotnet/winforms/tree/main/src/System.Windows.Forms/tests/IntegrationTests/DesignSurface/DemoConsole) test app in our repo, which I assume is similar to yours. And running it I can also reproduce the sizing issue you mentioned:
![DemoConsoleRepro](https://user-images.githubusercontent.com/102954094/213747452-a0f31eae-5adf-451f-bc96-cd2083f48e35.gif)

But looking at the gif from https://github.com/dotnet/winforms/pull/4860, where this functionality was introduced, changing size by mouse worked well. @Olina-Zhang - could you please take a look and create a new issue for being unable to change control's size by mouse in DemoConsole app? It reproduces both in .NET 7 and .NET 8, but works well in .NET 6.

In .NET 8 I also see another issue when trying to move controls on the surface:
![image](https://user-images.githubusercontent.com/102954094/213748563-ff65687e-f566-4624-aed8-5c67cce4be81.png) 
@Olina-Zhang - if this is not a known issue, can you create it for this bug as well?

**Comment by Olina-Zhang** [comment]
> @Olina-Zhang - if this is not a known issue, can you create it for this bug as well?

Opened GH issue: https://github.com/dotnet/winforms/issues/8515.

**Comment by dmanter** [comment]
> @dmanter thank you for a report, this seems to be a bug in .NET 7. We have the [DemoConsole](https://github.com/dotnet/winforms/tree/main/src/System.Windows.Forms/tests/IntegrationTests/DesignSurface/DemoConsole) test app in our repo, which I assume is similar to yours. And running it I can also reproduce the sizing issue you mentioned: ![DemoConsoleRepro](https://user-images.githubusercontent.com/102954094/213747452-a0f31eae-5adf-451f-bc96-cd2083f48e35.gif) [ ![DemoConsoleRepro](https://user-images.githubusercontent.com/102954094/213747452-a0f31eae-5adf-451f-bc96-cd2083f48e35.gif) ](https://user-images.githubusercontent.com/102954094/213747452-a0f31eae-5adf-451f-bc96-cd2083f48e35.gif) [ ](https://user-images.githubusercontent.com/102954094/213747452-a0f31eae-5adf-451f-bc96-cd2083f48e35.gif)
> 
> But looking at the gif from #4860, where this functionality was introduced, changing size by mouse worked well. @Olina-Zhang - could you please take a look and create a new issue for being unable to change control's size by mouse in DemoConsole app? It reproduces both in .NET 7 and .NET 8, but works well in .NET 6.
> 
> In .NET 8 I also see another issue when trying to move controls on the surface: ![image](https://user-images.githubusercontent.com/102954094/213748563-ff65687e-f566-4624-aed8-5c67cce4be81.png) @Olina-Zhang - if this is not a known issue, can you create it for this bug as well?
> 
> @dmanter - as a workaround, can you try to downgrade to .NET 6 and see if it resolves the issue with sizing?

@dmitrii-drobotov  - Yes my Form Designer on same idea as your Tiny Form Designer. I tried to do .Net6 but I have supporting dlls and use Telerik RadDock and those all target .net 7 so got lots of errors. For now I can resize as you did in the property editor manually so that gets me by until you all fix the sizing issue. I see you said above DateTimePicker could be done easy but not DataGridView. So please forgive me as the 6 lines of code given that @Leon99 supplied work for both. It seems such an easy fix but I certainly don't understand the coding at this deeper level so I do want to minimize the task. Now that this work around allows me to drop DTP and DGV onto forms everything seems to work so a bit confused. In your Tiny Form Designer in .Net 7 can you drop these 2 controls? Maybe I am doing something else wrong or at the very least do not understand the task. Maybe if you can do these two controls, then something in your Demo code is different than what I have and should look at your demo code? This project was done by another programmer years ago based on another VS Demo for a custom form designer which I believe cam from back a decade or so. Perhaps if these 2 controls work for you, there is new code I am missing. Thanks for your reply in advance!

**Comment by dmitrii-drobotov** [comment]
@dmanter - no, as of now DateTimePicker and DataGridView wouldn't work in our Tiny Form Designer either. Regarding DataGridView - if we port it in the very minimal way like @Leon99 suggested, it would lack some of the functionality that you see in VS designer for DataGridView, such as adding and editing columns, configuring cell styles, etc. So porting it completely might be complicated.

**Comment by dmanter** [comment]
> @dmanter - no, as of now DateTimePicker and DataGridView wouldn't work in our Tiny Form Designer either. Regarding DataGridView - if we port it in the very minimal way like @Leon99 suggested, it would lack some of the functionality that you see in VS designer for DataGridView, such as adding and editing columns, configuring cell styles, etc. So porting it completely might be complicated.

AHA - I don't need a complete port as my Designer allows a custom extension for specifying the Data Source as a SQL select statement. The datagrid is only ever populated at runtime by that select. As far as editing cols, there is a custom property for EditMode. That can be a list of CELLS(CellName1{Format};CellName2{Format}) to allow editing in the DGV or specified
 as FORM which will dynamically build an Edit Form that pops up to edit a Rows data. We've also in the DataGridView custom extension, removed lots of properties and just left what we allow to be edited. That all seems to work now with @Leon99 little fix.
![DGV](https://user-images.githubusercontent.com/3495842/214668089-2ff0d4b8-8154-423d-8a6d-f561a3c9af6f.JPG)

**Comment by dmanter** [comment]
Thanks @dmitrii-drobotov for your comments/support. As far as my needs, @Leon99 fix has resolved my issue. Thanks!

**Comment by dmanter** [comment]
OOPS- Forgot about the resizing handles not working so re-opened.

**Comment by elachlan** [comment]
@dmanter, resizing was fixed in #8522. I think a backport is required for .NET7.

**Comment by merriemcgaw** [comment]
@dmanter before we can service .NET 7 we need a bit of business justification. Are you able to wait until .NET 8 and if not, is this a total blocker for your application?

**Comment by Tanya-Solyanik** [comment]
Changed the title of this post because the DateTimePicker designer had been added already.
